python/datasources/cawp.py

```diff
@@ -0,0 +1,316 @@
+
+
+def get_standard_code_from_cawp_phrase(cawp_place_phrase: str):
+    """ Accepts a CAWP place phrase found in the LINE ITEM table
+    `{STATE_COL_LINE NAME} - {CODE}` with the standard 2 letter code
+     """
+
+    # swap out non-standard 2 letter codes
+    cawp_place_phrase = {"American Samoa - AM":
+                         "American Samoa - AS",
+                         "Northern Mariana Islands - MI":
+                         "Northern Mariana Islands - MP"}.get(
+                             cawp_place_phrase, cawp_place_phrase)
+
+    # only keep 2 letter code portion
+    place_terms_list = cawp_place_phrase.split(" - ")
+    place_code = place_terms_list[1]
+
+    return place_code
+
+
+def remove_markup(datum: str):
+    """Returns the string with any asterisks and/r italics markup removed """
+    datum = str(datum)
+    return datum.replace("<i>", "").replace("</i>", "").replace("*", "")
+
+
+def count_matching_rows(df, place_code: str, gov_level: str, race_to_match: str):
+    """ Accepts a dataframe, a level of government, a place code,
+    and a race name (CAWP terminology) string to match within the
+     race_ethnicity column. It then counts the number of
+    rows where those conditions are all met  """
+
+    df = df[(df[POSITION_COL].isin(CAWP_DATA_TYPES[gov_level]))]
+
+    # to get national values, don't restrict by state
+    if place_code != constants.US_ABBR:
+        df = df[(
+            df[POSTAL_COL] == place_code)]
+
+    # to get ALL women, don't restrict by race
+    if race_to_match == std_col.ALL_VALUE:
+        race_to_match = ""
+
+    # find race matches
+    df_race_matches = df[(df[RACE_COL].str.contains(race_to_match))]
+
+    if race_to_match != "Multiracial Alone":
+        return len(df_race_matches.index)
+
+    # combine 3 CAWP categorizations to form MULTI_OR_OTHER_STANDARD
+    # - "Multiracial Alone"
+    # - ", " (women who have a list of specific races)
+    # - "Other"
+    df_race_list_matches = df[(df[RACE_COL].str.contains(", "))]
+    df_race_other_matches = df[(df[RACE_COL].str.contains("Other"))]
+
+    return len(df_race_matches.index) + len(df_race_list_matches.index) + len(df_race_other_matches.index)
+
+
+CAWP_LINE_ITEMS_FILE = "cawp-by_race_and_ethnicity.csv"
+CAWP_TOTALS_URL = "https://cawp.rutgers.edu/tablefield/export/paragraph/1028/field_table/und/0"
+
+# Tables for US LEVEL LEG TOTALS
+PROPUB_US_SENATE_FILE = "propublica-us-senate.json"
+PROPUB_US_HOUSE_FILE = "propublica-us-house.json"
+
+
+class CAWPData(DataSource):
+
+    @ staticmethod
+    def get_id():
+        return 'CAWP_DATA'
+
+    @ staticmethod
+    def get_table_name():
+        return 'cawp_data'
+
+    def upload_to_gcs(self, _, **attrs):
+        raise NotImplementedError(
+            'upload_to_gcs should not be called for CAWPData')
+
+    def write_to_bq(self, dataset, gcs_bucket, **attrs):
+
+        # load in line-item table from CAWP with all women all levels by race/state
```

Comment:
Cool, good idea. It's a lot cleaner this way. Should i separate the "fetch" part from the data frame standardization part, and then add unit tests to the standardization part? Or is it ok to just leave the fn as FETCH+MANIPULATE and just rely on the integration tests ?

Reply:
I've been thinking about that too. Im not sure. I think its technically better practice to split it up and do simpler tests. But I do think that would lead to a bit of over testing and make it hard to change anything.

Reply:
But either way we can rely on the integration tests because everything needs to be under the `write_to_bq` function